rfd/0113-moderated-file-transfers.md

```diff
@@ -0,0 +1,176 @@
+
+---
+authors: Michael Myers (michael.myers@goteleport.com)
+state: draft
+---
+
+# RFD 113 - Web Moderated Session File Transfer
+
+## Required approvers
+
+Engineering: @zmb3 @xacrimon @jakule
+
+## What
+
+Approval process to allow file transfers during a web moderated session
+
+## Why
+
+Users have no ability to transfer files if their role requires a moderated session for the specified server, even if they are in an active and "valid" session currently. Originally, we would allow file transfers regardless of moderated sessions which, while good for UX, goes against the purpose of moderated sessions in the first place. We have disabled the functionality but would like to reintroduce it in a secure, moderated, and auditable way. 
+
+## Scope
+
+This RFD will cover how to implement for web SCP specifically, but may be expanded to SFTP for tsh, although my research has been mostly for web scp. If at anytime there is ambiguity in this RFD for "which type of file transfers does this refer to", default to web scp only. 
+
+## The current issue
+
+When a user initiates a file transfer request (HTTP request and `tcp scp`), the system establishes a new non-interactive session to run the "exec" command. When we attempt to [open the session](https://github.com/gravitational/teleport/blob/64b10f1ccbeeab63c4ce91f5f11fdd74b42448d8/lib/srv/sess.go#L277) we have a check named `checkIfStart` which will pull any policies that require moderation from the current identity, and return if they've been fulfilled or not. This will always fail as this is a "new" session, created just a few lines before the check itself. Therefore, we need a way to let the server know that this request has been moderator approved.
+
+### Details
+
+#### The FileTransferRequest 
+
+The general idea of the solution is to provide a way for users to "request-a-request" that can be approved by moderators. An example of a file transfer specific struct could look like
+
+```go
+type FileTransferRequest {
+    id string // UUID
+    user string
+    sessionID string
+    filePath string
+    approvers []*party
+    sink bool // upload or download
+}
+```
+
+We can add any relevant information otherwise needed. File-size was considered, but imagine a user trying to download a log file that could have changed size by the time the request flow has completed. 
+
+Because this struct is going to allow an exec command to go through, we could abstract it further to _any_ SSHRequest being passed instead with a struct like below
```

Comment:
If this was abstracted, this could allow access requests for a single command, instead of an interactive session (if that isn't already possible), would be interesting